src/ms_filter.jl

```diff
@@ -0,0 +1,209 @@
+    d::Int
+    m::Int 
+    n_input::Int
+end
+
+Adapt.adapt_structure(to, flt::ModifiedSincFilter) = flt
+
+function fltinstance(flt::ModifiedSincFilter, fi::SamplingInfo{T}) where {T}
+    _m = round(Int, ustrip(NoUnits, flt.m/step(fi.axis)))
+    m_min = flt.d/2 + 2
+    s2 = "size of kernel to small for given degree"
+    @assert _m >= m_min s2
+
+    ModifiedSincFilterInstance{T}(flt.d, _m, length(fi.axis))
+end
+
+RadiationDetectorDSP.flt_output_smpltype(fi::ModifiedSincFilterInstance) = RadiationDetectorDSP._floattype(flt_input_smpltype(fi))
+RadiationDetectorDSP.flt_input_smpltype(::ModifiedSincFilterInstance{T}) where {T} = T
+RadiationDetectorDSP.flt_output_length(fi::ModifiedSincFilterInstance) = fi.n_input
+RadiationDetectorDSP.flt_output_time_axis(::ModifiedSincFilterInstance, time::AbstractVector{<:RealQuantity}) = time
+
+function RadiationDetectorDSP.rdfilt!(output::AbstractVector, 
+    fi::ModifiedSincFilterInstance, input::AbstractVector)
+
+    @assert firstindex(input) == firstindex(output) "output and input signal don't have matching first indices"
+    @assert lastindex(input) == lastindex(output) "output and input signal don't have matching last indices"
+
+    T = eltype(output)
+    _y = similar(output, fi.n_input + fi.m*2)
+    _unsafe_extendData!(_y, input, fi.m, fi.d)
+    kernel = _makeKernel(fi.d, fi.m)
+    sum::T = zero(T)
+    @inbounds for i in eachindex(input)
+        center = fi.m + i
+        sum = fma(kernel[1], _y[center], sum)
+        @simd for j in Base.OneTo(length(kernel)-1)
+            k_j = kernel[j+1]
+            sum = fma(k_j, _y[center + j+1], sum)
+            sum = fma(k_j, _y[center - (j+1)], sum)
+        end
+        output[i] = sum
+        sum = zero(T)
+    end
+    output
+end
+
+const MS_COEFFS = Dict(
+    0 => Float64[],
+    2 => Float64[],
+    4 => Float64[],
+    6 => [0.001717576 0.02437382 1.64375],
+    8 => [0.0043993373 0.088211164 2.359375; 0.006146815  0.024715371 3.6359375],
+    10 => [0.0011840032 0.04219344 2.746875; 0.0036718843 0.12780383 2.7703125]
+)
+
+@inbounds function _makeKernel(d::Int, m::Int)
+    coeffs = MS_COEFFS[d]
+    T = RadiationDetectorDSP._floattype(Int)
+    kernel = Vector{T}(undef, m+1)
+    sincArg = (d+4)/2
+    sum::T = zero(T)
+    κ::Vector{T} = map(n -> κ_j(m, view(coeffs, n, 1:3)), axes(coeffs, 1))
+    ν = isodd(d/2) ? 1 : 2
+    α = 4
+    for i in eachindex(kernel)
+        x = (i - 1)/(m + 1)
+        k_i = sinc(sincArg*x)
+        @simd for j in eachindex(κ)
+            s = x*sin((2*j + ν)*π*x)
+            k_i = fma(κ[j], s, k_i)
+        end
+        k_i *= _w(x, α)
+        kernel[i] = k_i
+        sum += k_i
+        i > 1 && (sum += k_i)   # offcenter kernel elements appear twice
+    end
+    kernel ./= sum
+    return kernel
+end
+
+"""
+    makeFitWeights(d::Int, m::Int)
+
+return the weights for the linear fit user for linear extrapolation at 
+at the right boundary.
+"""
+function makeFitWeights(d::U, m::Int) where U
```

Comment:
Looks like `d` can be anything according to the function but should be an `Int` according to the docstring. Does it make sense to constrain `U` to be `Real`, `Number` or `Integer`?